STAR.py

```diff
@@ -1,11 +1,145 @@
+    """
+        Given an election instance and optionally specified tiebreakers, return the STAR winner.
+
+        Parameters:
+            ballots (pd.DataFrame): contains a dataframe oriented such that each column name is a candidate
+                and each row is a voter's ballot. That is, the score voter A gives to candidate X is ballots[X][A]
+            
+            scoring_tiebreaker (function): Given an election instance, returns a winner according to the tiebreaker
+                or a TrueTie instance if it is not sufficiently resolute. Called during the scoring round of STAR.
+            
+            runoff_tiebreaker (function): Given an election instance, returns a winner according to the tiebreaker
+                or a TrueTie instance if it is not sufficiently resolute. Called during the runoff round of STAR.
+    """
+
+    if ballots.shape[1] < 1:
+        raise InvalidElection("Not enough candidates to fill desired number of seats")
+
+    # If there is only one candidate, elect them
+    if ballots.shape[1] == 1:
+        return ballots.columns[0]
+
+    runoff_candidates = []
+    while len(runoff_candidates) < 2:
+
+        eligible = ballots.drop(runoff_candidates, axis=1)
+        w = scoring_tiebreaker(ballots[score_winners(eligible)])
+        if isinstance(w, TrueTie):
+
+            if len(w.tied) == 2 and len(runoff_candidates) == 0:
+                runoff_candidates.extend(w.tied)
+
+            else:
+                # In this case, the election returned a tie unresolvable by stated tiebreakers
+                w.tied.extend(runoff_candidates)
+                return w
+
+        else:
+            runoff_candidates.append(w)
     
-    #Run off
-    winner = df_tops.eq(df_tops.max(1), axis=0).sum().idxmax()  
+    # At this point, either we have already exited or runoff_candidates contains exactly two candidates
+    a,b = runoff_candidates
+    runoff_outcome = pairwise_winner(ballots[a], ballots[b])
+
+    if runoff_outcome == 1:
+        return a
+    elif runoff_outcome == -1:
+        return b
+    else:
+        return runoff_tiebreaker(ballots[runoff_candidates])
+
+def Bloc_STAR(ballots: pd.DataFrame, numwinners: int, scoring_tiebreaker=default_scoring_tiebreaker, runoff_tiebreaker=default_runoff_tiebreaker):
+    elected = []
+    while len(elected) < numwinners:
+        
+        eligible = ballots.drop(elected, axis=1)
+        w = STAR(eligible, scoring_tiebreaker=scoring_tiebreaker, runoff_tiebreaker=runoff_tiebreaker)
+        if isinstance(w, TrueTie):
+
+            if len(w.tied) + len(elected) <= numwinners:
+                # If multiple candidates are tied in STAR but we have enough seats to elect them all, do so
+                elected.extend(w.tied)
+
+            else:
+                # In this case, the election returned a tie unresolveable by stated tiebreakers
+                w.tied.extend(elected)
```

Comment:
Let's say I run a 5 winner BLOC election, and there are 4 winners + 2 people tied for the 5th spot. This system would output that as a 6-way tie, is that correct? Would it be more accurate to output 4 winners, and 2 more tied candidates?

I'm not sure if there's a better approach without making the output more involved. I just want to make sure I'm understanding it correctly

Reply:
Yes that's correct. The other way is reasonable too, but in the end I had to pick one of the paradigms. If you'd like that changed let me know. Personally, I don't think of it as a 6-way tie, more like just a flag that the election could not resolve to 6 winners.

Reply:
One simple way to distinguish the elected cands from the tied ones is to add an attribute `elected` to `TrueTie` and extend that with the elected candidates rather than `tied`

Reply:
Cool, yeah I think I like that approach better, but I'm fine either way

Reply:
(now that you've made me think about it 🙂 ) I like that approach better as well. updated below